### DISCUSSION TITLE
Add hover to vertices and be able to remove vertex by click on hover icon

### DISCUSSION THREAD

**Discussion by AlexBezmoshchuk** [clientIO/joint#2559] [Q&A] [answered]
### Introduction

Initially I used JointJS 3.7.5 with fsa.Arrow as links between elements. They had pretty useful feature to remove vertices by clicking to close button in hover.
![image](https://github.com/clientIO/joint/assets/122964943/88b9be82-926f-4917-9ba7-1ba3080753e8)
In version 4 fsa package is removed, so I replaced my links by using custom links (defined by shapes.standard.Link) and they work great, but I lost a feature to remove vertices by clicking on hover. Double click event is used by another feature, so I turned off vertexRemoving. 

Is it possible to add hover to vertices or improve markup of vertices to create element with hover tooltip? I tried to extend VertexHandle using

```
const customVertices = new linkTools.Vertices({
  handleClass: linkTools.Vertices.VertexHandle.extend({
    options: { markup: util.svg`<rect fill="red" width="10" height="20" x="-10" y="-10"/>` }, // some test markup to see if it works
  }),
  vertexAdding: false,
  vertexRemoving: false,
});
```
but vertex shape is the same

### Restrictions & Constraints

- Cannot use standard way to remove vertices (by double click on them) due to business restrictions.

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS

**Comment by kumilingus** [comment]
The `VertexHandle` does not read `markup` option. You can do it this way:
```js
const customVertices = new linkTools.Vertices({
    handleClass: linkTools.Vertices.VertexHandle.extend({
        tagName: 'g',
        children: util.svg`<rect fill="red" width="10" height="20" x="-10" y="-10"/>`,
        onRender() {
            this.renderChildren();
        }
    }),
    vertexAdding: false,
    vertexRemoving: false,
});
```

**Reply by AlexBezmoshchuk** [reply]
Thanks a lot. I changed the markup to have a standard shape of vertices with remove tooltip.
```
const handleClick = (event: MouseEvent) => {
    if (!event.currentTarget) {
      return;
    }

    const lineElement = event.currentTarget.parentElement.parentElement;
    const [x, y] = lineElement.getAttribute('transform').replace('matrix(1,0,0,1,', '').replace(')', '').split(',');
    const linkId = lineElement.getAttribute('model-id');

    onRemove(linkId, x, y);
};

return new linkTools.Vertices({
    vertexAdding: false,
    vertexRemoving: false,
    handleClass: linkTools.Vertices.VertexHandle.extend({
      tagName: 'g',
      children: util.svg`
<circle r="6" stroke="#FFFFFF" stroke-width="2" cursor="move"></circle>
<g class="vertex-tooltip__remove">
        <g class="vertex-tooltip__remove-btn" cursor="pointer">
            <circle r="5" stroke="red" fill="red" stroke-width="2" />
            <line x1="-3" y1="-3" x2="3" y2="3" stroke="#fff" stroke-width="2" />
            <line x1="3" y1="-3" x2="-3" y2="3" stroke="#fff" stroke-width="2" />
        </g>
    </g>
`,
      onRender() {
        this.renderChildren();

        const removeBtns = this.el.querySelectorAll('.vertex-tooltip__remove-btn');
        removeBtns.forEach((btn: HTMLElement) => {
          btn.addEventListener('click', handleClick);
        });
      },

      onRemove() {
        const removeBtns = this.el.querySelectorAll('.vertex-tooltip__remove-btn');
        removeBtns.forEach((btn: HTMLElement) => {
          btn.removeEventListener('click', handleClick);
        });
      },
    }),
});
```

The only concern is I haven't found a way to get a vertex properties such as lineId, x and y without getting it from line element (with class .joint-marker-vertex). Is there a way to get such data from `this` in method onRender?

**Reply by kumilingus** [reply, marked as answer]
It's easier to use the same mechanisms that were used in the original `VertexHandle` class.
Would this work? https://jsfiddle.net/kumilingus/ke60L73q/

**Reply by AlexBezmoshchuk** [reply]
This is great. Thanks a lot.
